src/libraries/System.Net.Http/src/System/Net/Http/MetricsHandler.cs

```diff
@@ -0,0 +1,235 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Net.Http.Metrics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http;
+
+internal sealed class MetricsHandler : HttpMessageHandlerStage, IHttpMetricsLogger
+{
+    private readonly HttpMessageHandler _innerHandler;
+    private readonly UpDownCounter<long> _currentRequests;
+    private readonly Counter<long> _failedRequests;
+    private readonly Histogram<double> _requestsDuration;
+
+    [ThreadStatic]
+    private static HttpMetricsEnrichmentContext? s_cachedEnrichmentContext;
+
+    public MetricsHandler(HttpMessageHandler innerHandler, Meter meter)
+    {
+        _innerHandler = innerHandler;
+
+        // Meter has a cache for
+        _currentRequests = meter.CreateUpDownCounter<long>(
+            "http-client-current-requests",
+            description: "Number of outbound HTTP requests that are currently active on the client.");
+        _failedRequests = meter.CreateCounter<long>(
+            "http-client-failed-requests",
+            description: "Number of outbound HTTP requests that have failed.");
+        _requestsDuration = meter.CreateHistogram<double>(
+            "http-client-request-duration",
+            unit: "s",
+            description: "The duration of outbound HTTP requests.");
+    }
+
+    public static Meter DefaultMeter { get; } = new SharedMeter();
+
+    internal override ValueTask<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
+    {
+        if (_currentRequests.Enabled || _failedRequests.Enabled || _requestsDuration.Enabled)
+        {
+            ArgumentNullException.ThrowIfNull(request);
```

Comment:
Why is this needed?

Reply:
`DiagnosticsHandler` has a [similar guard](https://github.com/dotnet/runtime/blob/318c0e6708ced35180fd5218170f82246e2f2bac/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L81), most likely because `HttpClientHandler` doesn't null-check the request in its' `SendAsync` overload before calling the next handler in the chain, while it's [expected to throw on null](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.sendasync?view=net-7.0#exceptions). We can remove the null-check from both `MetricsHandler` and `DiagnosticsHandler` and add it to `HttpClientHandler` instead.

Reply:
> most likely because HttpClientHandler doesn't null-check the request in its' SendAsync overload before calling the next handler in the chain

Can you elaborate? HttpClientHandler.SendAsync just calls `Handler.SendAsync`:
https://github.com/dotnet/runtime/blob/e5ddc6c9b7e481d36d6350960868188d7b275ef7/src/libraries/System.Net.Http/src/System/Net/Http/HttpClientHandler.cs#L309-L310
and `Handler` is just the underlying handler, which is a SocketsHttpHandler, and SocketsHttpHandler.SendAsync does null check:
https://github.com/dotnet/runtime/blob/e5ddc6c9b7e481d36d6350960868188d7b275ef7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L557

Reply:
> and `Handler` is just the underlying handler, which is a SocketsHttpHandler

On mobile and browser it is `DiagnosticsHandler` on main, unless `DiagnosticsHandler.IsGloballyEnabled() == false`. With the PR's change, `MetricsHandler` is going to be the first handler on those platforms.